templates/openstackconfiggenerator/config/16.2/rendered-tripleo-config.yaml

```diff
@@ -28,6 +28,17 @@ resource_registry:
 {{- end }}
 {{- end }}
 
+# Parameter merge strategies
+parameter_merge_strategies:
```

Comment:
@rabi could we just set the default to merge for 16.x?

Reply:
Looking at the heat merge code that wouldn't work - it would join string params and add number params!

Reply:
I don't think it's required for 16/17. It's not documented that maps would be merged in anyway. In case someone one wants to change these parameters, they are expected to provide full ServiceNetMap, EndpointMap[1] etc and the ones in the last environment would override.

However, we've changed some of those in master where we specify merge strategies[2]. parameter_merge_strategy: merge for a map/list won't do a deep merge (as you mentioned), just a simple merge[3]. 

[1] https://github.com/openstack/tripleo-heat-templates/blob/stable/train/environments/ssl/tls-endpoints-public-dns.yaml#L14 
[2] https://github.com/openstack/tripleo-heat-templates/blob/master/overcloud-resource-registry-puppet.j2.yaml#L345
[3] https://github.com/openstack/heat/blob/master/heat/common/environment_util.py#L58

Reply:
STF is relying on this undocumented merge behaviour for ExtraConfig

Reply:
I've not seen any issues reported for OSP17 yet though.  I guess would expect them to change instead and not depend on undocumented behavior though:)

Reply:
Probably just not hit this for 17 yet :-)

E.g this is relying on ServiceNetMap merging: https://github.com/tungstenfabric/tf-tripleo-heat-templates/blob/stable/train/environments/contrail/contrail-services.yaml#L2
And for STF the docs say to include the qdr-edge-only.yaml environment file (which sets ExtraConfig) and then creating a custom environment file that also sets ExtraConfig.

Reply:
Trilio too https://github.com/trilioData/triliovault-cfg-scripts/blob/master/redhat-director-scripts/rhosp16.2/environments/trilio_env.yaml#L17

Reply:
That would be a STF bug (not TripleO), if reported ;) I only changed in xena[1] in THT. Probably we should not change something that's not inline documentation(for 17). 

[1] https://github.com/openstack/tripleo-heat-templates/commit/b253d564f7cccd0a818239a3c877229242acfe55

Reply:
I don't see any difference in 16.2/17 for director-operator, as we're using stack api[1] and not overcloud deploy (which merges environment on the client side for 16.2). Not sure what changing for 16.2 only would achieve. 

Also, as I mentioned earlier just adding here won't help. It has to be added to all environment files where these parameters are mentioned.

[1] https://github.com/openstack-k8s-operators/osp-director-operator/blob/master/templates/openstackconfiggenerator/bin/create-playbooks.sh#L88

Reply:
Yes it's per file and for a reason[1].

"You can provide a default merge strategy and/or parameter specific merge strategies per environment file"

[1] https://docs.openstack.org/heat/latest/template_guide/environment.html#environment-merging